### PR TITLE
Test: Add unit and integration tests for `totalBalanceOf()` and pool manager transfer functions

### LINKED ISSUE #756: TotalBalanceOf unit & integration tests
Write unit & integration tests for the totalBalanceOf (with and without delegations) from this PR: https://github.com/NexusMutual/smart-contracts/pull/753/

Open a separate PR based on this branch ^

### PR DESCRIPTION
**Pull request by kyledewy**
## Context

closes #756 

## Changes proposed in this pull request
Added tests for the changes made [here](https://github.com/NexusMutual/smart-contracts/pull/753)

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [x] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: contracts/modules/token/TokenController.sol [hunk 1]
```diff
@@ -171,6 +171,10 @@ contract TokenController is ITokenController, LockHandler, MasterAwareV2 {
         continue;
       }
       uint256 amount = locked[users[i]]["CLA"].amount;
+
+      // TODO: test this function
+      assert (amount == 0);
```
Comment (kyledewy) [inline contracts/modules/token/TokenController.sol:176]: TODO: remove assert

Comment (kyledewy) [reply contracts/modules/token/TokenController.sol:176]: I don't see a method of setting the `locked` storage value. I tried to investigate where it gets set, but it seems none of our tests do. Is this fully implemented?

File: test/unit/TokenController/totalBalanceOf.js [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    const amount = One.mul(2).pow(80);
+    await assessment.setStakeOf(member1.address, amount);
+
+    expect(await tokenController.totalBalanceOf(member1.address)).to.equal(amount.add(this.nxmBalanceBefore));
+    expect(await tokenController.totalBalanceOfWithoutDelegations(member1.address)).to.equal(
+      await tokenController.totalBalanceOf(member1.address),
+    );
+  });
+
+  async function stakingPoolAddressAt(poolFactoryAddress, poolId) {
```
Comment (shark0der) [inline test/unit/TokenController/totalBalanceOf.js:79]: Move up before the describe

File: test/unit/TokenController/totalBalanceOf.js [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    const { assessment, tokenController } = this.contracts;
+    const [member1] = this.accounts.members;
+
+    const amount = One.mul(2).pow(80);
+    await assessment.setStakeOf(member1.address, amount);
+
+    expect(await tokenController.totalBalanceOf(member1.address)).to.equal(amount.add(this.nxmBalanceBefore));
+    expect(await tokenController.totalBalanceOfWithoutDelegations(member1.address)).to.equal(
+      await tokenController.totalBalanceOf(member1.address),
+    );
```
Comment (shark0der) [inline test/unit/TokenController/totalBalanceOf.js:76]:
```
    const amount = One.mul(2).pow(80);
    await assessment.setStakeOf(member1.address, amount);

    const expecedAmount = this.nxmBalanceBefore.add(amount);

    expect(await tokenController.totalBalanceOf(member1.address)).to.equal(expecedAmount);
    expect(await tokenController.totalBalanceOfWithoutDelegations(member1.address)).to.equal(expecedAmount);
```

File: test/unit/TokenController/totalBalanceOf.js [hunk 4]
```diff
@@ -0,0 +1,155 @@
+const { expect } = require('chai');
+const { ethers } = require('hardhat');
+const { setEtherBalance } = require('../../utils/evm');
+const { One, MaxUint256 } = ethers.constants;
+
+const poolId = 234;
+describe('totalBalanceOf', function () {
+  beforeEach(async function () {
+    const { nxm } = this.contracts;
+    const [member1] = this.accounts.members;
```
Comment (shark0der) [inline test/unit/TokenController/totalBalanceOf.js:10]: Since it's a single member that you're using - name it `member` instead of `member1`.

File: test/unit/TokenController/totalBalanceOf.js (hunk 3 above)
Comment (shark0der) [reply test/unit/TokenController/totalBalanceOf.js:76]: This pattern seems to be repeated in a few places. The suggested change is easier to read.

File: contracts/modules/token/TokenController.sol (hunk 1 above)
Comment (shark0der) [reply contracts/modules/token/TokenController.sol:176]: It was implemented, it's functionality that was removed from the contracts. We'll test it in fork tests, skip unit tests for this part.

File: package.json [hunk 5]
```diff
@@ -58,6 +58,7 @@
     "eslint-plugin-standard": "^5.0.0",
     "hardhat": "^2.11.2",
     "hardhat-contract-sizer": "^2.6.1",
+    "hardhat-tracer": "^2.0.0",
```
Comment (shark0der) [inline package.json:61]: Please extract this as a separate PR. I think you included this in another PR as well.

**Comment by kyledewy** [comment]
Coverage is failing when impersonating a StakingPool. I initially thought it may be because there is no code at the address, but that didn't seem to help.

File: test/unit/TokenController/createStakingPoolOwnershipOffer.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+      .connect(oldManager)
+      .createStakingPoolOwnershipOffer(poolId, newManager2.address, expectedDeadline2);
+
+    const { proposedManager, deadline } = await tokenController.getStakingPoolOwnershipOffer(poolId);
+    expect(proposedManager).to.be.equal(newManager2.address);
+    expect(deadline).to.be.equal(expectedDeadline2);
+  });
+
+  // TODO: not sure if we want this behavior?
+  it('should successfully transfer ownership to the same address', async function () {
```
Comment (roxdanila) [inline test/unit/TokenController/createStakingPoolOwnershipOffer.js:85]: should be pretty much no-op

File: test/unit/TokenController/totalBalanceOf.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    expect(await tokenController.totalBalanceOf(member.address)).to.equal(
+      delegateAmount.add(stakerDepositAmount).add(stakerRewardAmount).add(mintedAmount).add(this.nxmBalanceBefore),
+    );
+    expect(await tokenController.totalBalanceOfWithoutDelegations(member.address)).to.equal(
+      (await tokenController.totalBalanceOf(member.address)).sub(delegateAmount),
+    );
+  });
+
+  // TODO: move to fork test
+  it.skip('should correctly calculate locked tokens', async function () {});
```
Comment (roxdanila) [inline test/unit/TokenController/totalBalanceOf.js:156]: we have this on the fork currently, it's fine to remove from here

File: test/unit/TokenController/transferStakingPoolOwnership.js [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    expect(poolsNewManager.length).to.be.equal(20);
+    expect(await tokenController.isStakingPoolManager(newManager.address)).to.be.equal(true);
+
+    // Check old manager
+    const poolsOldManager = await tokenController.getManagerStakingPools(oldManager.address);
+    expect(poolsOldManager.length).to.be.equal(0);
+    expect(await tokenController.isStakingPoolManager(oldManager.address)).to.be.equal(false);
+  });
+
+  // TODO: not sure if we want this behavior
```
Comment (roxdanila) [inline test/unit/TokenController/transferStakingPoolOwnership.js:92]: given we only use it within switchMembership now, this is fine as the new address becomes a member.
however, switchMembership should check that the `to` address cannot be 0. I'll create a separate issue for this

**Comment by roxdanila** [comment]
> Coverage is failing when impersonating a StakingPool. I initially thought it may be because there is no code at the address, but that didn't seem to help.

You should be doing something similar to this: https://github.com/NexusMutual/smart-contracts/blob/nexus-v2/test/unit/Cover/createStakingPool.js#L53

**Review by roxdanila — CHANGES_REQUESTED** [review]
left a suggestion on how to fix the coverage job, and comments regarding the leftover todos

**Review by roxdanila — APPROVED** [review]
